Add graphics package to existing switches

dra27 opened the pull request:
At present, the `graphics` package allows the absence of the OCaml graphics library to be detected as a dependency failure (which is neater than a build system error). This PR enhances that by building and installing the graphics library if it's missing. Here's an example:

<pre><font color="#8AE234"><b>dra@bionic</b></font>:<font color="#729FCF"><b>~</b></font>$ ocamlfind query graphics
ocamlfind: Package `graphics&apos; not found

<font color="#8AE234"><b>dra@bionic</b></font>:<font color="#729FCF"><b>~</b></font>$ opam install graphics
The following actions will be performed:
  <font color="#4E9A06">∗</font> install <b>graphics</b> 1.0

<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>Gathering sources</b> <font color="#06989A">&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>

<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>Processing actions</b> <font color="#06989A">&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>
<font color="#CC0000">[ERROR]</font> The compilation of graphics failed at &quot;/home/dra/.opam/opam-init/hooks/sandbox.sh build ocamlc -custom graphics.cma -o test&quot;.

<font color="#CC0000">#===</font> <b>ERROR</b> <b>while compiling graphics.1.0</b> <font color="#CC0000">=======================================#</font>
...
<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>Error report</b> <font color="#06989A">&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>
<font color="#CC0000">┌─ </font>The following actions <font color="#CC0000">failed</font>
<font color="#CC0000">│ </font><font color="#06989A">λ</font> build <b>graphics</b> 1.0
<font color="#CC0000">└─ </font>
<font color="#06989A">╶─ </font>No changes have been performed

<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>graphics.1.0 troubleshooting</b> <font color="#06989A">&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>
<font color="#CC0000">=&gt; </font>This package checks whether the Graphics library was compiled.

<font color="#8AE234"><b>dra@bionic</b></font>:<font color="#729FCF"><b>~</b></font>$ opam remote add graphics git+https://github.com/dra27/opam-repository.git#graphics
...

<font color="#8AE234"><b>dra@bionic</b></font>:<font color="#729FCF"><b>~</b></font>$ opam install graphics --yes
The following actions will be performed:
  <font color="#4E9A06">∗</font> install <b>conf-pkg-config</b> 1.1    [required by conf-libX11]
  <font color="#4E9A06">∗</font> install <b>conf-libX11</b>     1      [required by graphics]
  <font color="#4E9A06">∗</font> install <b>graphics</b>        4.06.1
===== <font color="#4E9A06">∗</font> 3 =====

<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>Gathering sources</b> <font color="#06989A">&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>
[<font color="#4E9A06">graphics.4.06.1</font>] found in cache

<font color="#06989A">&lt;&gt;&lt;&gt;</font> <b>Processing actions</b> <font color="#06989A">&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;</font>
<font color="#4E9A06">∗</font> installed <b>conf-pkg-config</b>.1.1
<font color="#4E9A06">∗</font> installed <b>conf-libX11</b>.1
<font color="#4E9A06">∗</font> installed <b>graphics</b>.4.06.1
Done.

<font color="#8AE234"><b>dra@bionic</b></font>:<font color="#729FCF"><b>~</b></font>$ ocamlfind query graphics
/home/dra/.opam/ocaml-with-graphics/lib/graphics
</pre>

For system switches, the library is installed as a findlib package only so, just as with nums in 4.06.0 onwards, users of system switches will need to be using something more sophisticated than `#load "graphics.cma"`. For opam-built compilers, they are installed to the OCaml libdir as OCaml would - the package installs a META file if ocamlfind has already been installed (ocamlfind will automatically generate META on any future installations).

A `conf-libX11` package is added which at the moment is based entirely on pkg-config. Part of removing `graph` and `win32graph` from OCaml may include transferring the `configure` logic to this package, but for now I kept it simple.

Reviewers may wish to note that the `install.sh` is identical in all packages and the `opam` file differs only by version-specific patches. Originally I did this all in `opam` files, but it's *really* difficult to deal with both system and non-system compilers **and** the fact that graphics may be already installed in the commands, so a script seemed somewhat less unclear.

These packages were generated by the scripts in https://github.com/dra27/opam-repository/commit/14f9326e9abec6ecada18a47ab55e7d2e65e42d9.

camelus commented:
##### :sunny: All lint checks passed <small>0d9310d0ae4ac3f121475b1b2c144328abbdd204</small>

* These packages passed lint tests: `conf-libX11.1`, `graphics.3.07+1`, `graphics.3.07+2`, `graphics.3.07`, `graphics.3.08.0`, `graphics.3.08.1`, `graphics.3.08.2`, `graphics.3.08.3`, `graphics.3.08.4`, `graphics.3.09.0`, `graphics.3.09.1`, `graphics.3.09.2`, `graphics.3.09.3`, `graphics.3.10.0`, `graphics.3.10.1`, `graphics.3.10.2`, `graphics.3.11.0`, `graphics.3.11.1`, `graphics.3.11.2`, `graphics.3.12.0`, `graphics.3.12.1`, `graphics.4.00.0`, `graphics.4.00.1`, `graphics.4.01.0`, `graphics.4.02.0`, `graphics.4.02.1`, `graphics.4.02.2`, `graphics.4.02.3`, `graphics.4.03.0`, `graphics.4.04.0`, `graphics.4.04.1`, `graphics.4.04.2`, `graphics.4.05.0`, `graphics.4.06.0`, `graphics.4.06.1`, `graphics.4.07.0`


---


##### :sun_behind_small_cloud: Installability check (9688 &rarr; 9722)

* new installable packages (34): `conf-libX11.1` `graphics.3.07` `graphics.3.08.0` `graphics.3.08.1` `graphics.3.08.2` `graphics.3.08.3`
  `graphics.3.08.4` `graphics.3.09.0` `graphics.3.09.1` `graphics.3.09.2` `graphics.3.09.3` `graphics.3.10.0` `graphics.3.10.1`
  `graphics.3.10.2` `graphics.3.11.0` `graphics.3.11.1` `graphics.3.11.2` `graphics.3.12.0` `graphics.3.12.1` `graphics.4.00.0`
  `graphics.4.00.1` `graphics.4.01.0` `graphics.4.02.0` `graphics.4.02.1` `graphics.4.02.2` `graphics.4.02.3` `graphics.4.03.0`
  `graphics.4.04.0` `graphics.4.04.1` `graphics.4.04.2` `graphics.4.05.0` `graphics.4.06.0` `graphics.4.06.1` `graphics.4.07.0`
* new **broken** packages (2): `graphics.3.07+1` `graphics.3.07+2`

dra27 commented:
The Camelus broken package is because this PR assumes #12669.

#12451 fixed the checksum for the clang patch in 4.00.1, but for some reason skipped 4.01.0, so that's bundled in here.

I've added stuff which seems to fix pkg-config detection of X11 on macOS!

I'm at a loss as to how to make DataKitCI happy - the first time it installed `conf-pkg-config` and `conf-libX11` and complained at `./install.sh` in the graphics package. So I changed it to be `sh ./install.sh` but then DataKitCI the second time suddenly can't install `conf-pkg-config`. Meh.

kit-ty-kate commented:
```
#=== ERROR while compiling graphics.4.07.0 ====================================#
"./install.sh": command not found.
```
I guess there is a `chmod +x` missing in this file

kit-ty-kate commented:
I don't understand the problem.. :/ The file already has the right permissions.. Anyone has any idea?

dra27 commented:
I tried a test commit changing the call to be `sh ./install.sh` but then I got errors about being unable to install `conf-pkg-config` instead :-/

dra27 commented:
Basically, there are various things which could be going on (error in opam’s logic to find the command, error in the script, etc.), but I gave up when it appeared that the process wasn’t stable.

dra27 commented:
@samoht - any ideas? avsm I know is way too snowed under to be able to look at this atm (of course, you may be too!)

dra27 commented:
I think I may have solved this - please don't merge, even if CI passes this time round

kit-ty-kate commented:
Ahhh
```
[WARNING] Could not read file /home/opam/.opam/repo/default/packages/graphics/graphics.4.07.1/opam. skipping:
          At /home/opam/.opam/repo/default/packages/graphics/graphics.4.07.1/opam:30:2:
          Parse error
```
Somehow it's a silent error :/

kit-ty-kate commented:
I've pushed a fix to get the error sooner next time: https://github.com/avsm/mirage-ci/commit/5d761efd83cb956a5175834b0fb5a3c76177df6c

dra27 commented:
Thanks, @kit-ty-kate! The log now [reveals the problem](https://ci.ocamllabs.io/log/saved/docker-build-0cff735fe6e4d7fbd5d19dc003d4d754/d22aad478c13be2426ab38afdd191be35deec1df) as well:

```
Processing  5/6: [graphics: ls install.sh]
+ /bin/ls "-l" "install.sh" (CWD=/home/opam/.opam/4.07/.opam-switch/build/graphics.4.07.1)
- -rw-r--r-- 1 opam opam 1980 Nov  1 18:57 install.sh
Processing  5/6: [graphics: ocaml unix.cma]
+ /home/opam/.opam/4.07/bin/ocaml "unix.cma" "probe.ml" (CWD=/home/opam/.opam/4.07/.opam-switch/build/graphics.4.07.1)
- uid=1000  groups=  stat=1000,1000,644

#=== ERROR while compiling graphics.4.07.1 ====================================#
"./install.sh": command not found.
```

The error message from opam should be better (https://github.com/ocaml/opam/issues/3649), but why is the file system not correct for DataKit CI?

dra27 commented:
OK, not sure how much more poking around I can sensibly do: https://ci.ocamllabs.io/log/saved/docker-build-2b6c0b6fad0045aab224e73d3691e9d1/e4aa5092ef8815c8d2ddad5b249ea940689f19ce shows that the `install.sh` files are definitely checked out from git with mode 755 but by the time they're ending up in the build directory they've lost them. Is one of the intermediate container snapshots not preserving execute bits for some reason?

kit-ty-kate commented:
This looks ok. Should I go ahead and merge it?

dra27 commented:
Huzzah - finally green! Please do 🙂

kit-ty-kate commented:
Thanks a lot!!
